### ISSUE TITLE
Question about seed

### ISSUE THREAD

**Issue by sorrowyn** [tfzhou/ProtoSeg#11]
```python
if args_parser.seed is not None:
	random.seed(args_parser.seed)
	torch.manual_seed(args_parser.seed)
```
Each gpu is set to the same seed.

```python
# fix the seed for reproducibility
if args_parser.seed is not None:
	from lib.utils.distributed import get_rank()
	seed = args_parser.seed + get_rank()
	torch.manual_seed(seed)
	np.random.seed(seed)
	random.seed(seed)
```
[Reference](https://github.com/IDEA-Research/DAB-DETR/blob/main/main.py)

**Comment by tfzhou** [comment]
Hi @sorrowyn Thanks for pointing this out, but it is a bit unclear to me why should we set different seeds for different gpus?

**Comment by sorrowyn** [comment]
https://zhuanlan.zhihu.com/p/250471767

> 1. DDP的N个进程都使用同一个随机数种子
> 2. 在生成数据时，如果我们使用了一些随机过程的数据扩充方法，那么，各个进程生成的数据会带有一定的同态性。比如说，YOLOv5会使用mosaic数据增强（从数据集中随机采样3张图像与当前的拼在一起，组成一张里面有4张小图的大图）。这样，因为各卡使用了相同的随机数种子，你会发现，各卡生成的图像中，除了原本的那张小图，其他三张小图都是一模一样的！
> 3. 同态性的数据，降低了训练数据的质量，也就降低了训练效率！最终得到的模型性能，很有可能是比原来更低的。

```python
import random
import numpy as np
import torch

def init_seeds(seed=0, cuda_deterministic=True):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    # Speed-reproducibility tradeoff https://pytorch.org/docs/stable/notes/randomness.html
    if cuda_deterministic:  # slower, more reproducible
        cudnn.deterministic = True
        cudnn.benchmark = False
    else:  # faster, less reproducible
        cudnn.deterministic = False
        cudnn.benchmark = True
        
def main():
    rank = torch.distributed.get_rank()
    # 问题完美解决！
    init_seeds(1 + rank)
```

**Comment by tfzhou** [comment]
Hi @sorrowyn, thanks for sharing this!
